File: 2022/rust/src/day1.rs
use std::{collections::BinaryHeap, num::ParseIntError};

pub fn run() {
    let mut calories = Calories::new();
    let input = include_str!("inputs/input1.txt");
    if let Err(e) = calories.parse(input) {
        println!("Error Parsing: {}", e);
    }
    println!("day1:");
    println!("part1: {}", calories.part1());
    println!("part2: {}", calories.part2());
}

#[derive(Default)]
struct Calories {
    items: BinaryHeap<i32>,
}

impl Calories {
    fn new() -> Self {
        Default::default()
    }

    fn parse(&mut self, input: &str) -> Result<(), ParseIntError> {
        let lines = input.split("\n\n");
        for lines_split in lines {
            let mut total = 0;
            for line in lines_split.lines() {
                total += line.parse::<i32>()?;
            }
            self.items.push(total);
        }
        Ok(())
    }

    fn part1(&self) -> i32 {
        *self.items.peek().unwrap()
    }

    fn part2(&mut self) -> i32 {
        let mut total = 0;
        for _ in 0..3 {
            total += self.items.pop().unwrap();
        }
        total
    }
}

File: 2021/day1/day1.go
package main

import (
	_ "embed"
	"fmt"
	"strconv"
	"strings"
)

//go:embed input.txt
var input string

type Report struct {
	depths []int
}

func (r *Report) Parse(v string) error {
	lines := strings.Split(v, "\n")
	r.depths = make([]int, len(lines))
	for i, str := range lines {
		if str == "" {
			continue // Skip empty lines
		}

		val, err := strconv.Atoi(str)

		if err != nil {
			return err
		}

		r.depths[i] = val
	}

	return nil
}

func (r *Report) slidingWindow() []int {
	var result []int
	for i := 0; i < len(r.depths)-2; i++ {
		sum := r.depths[i] + r.depths[i+1] + r.depths[i+2]
		result = append(result, sum)
	}
	return result
}

func (r *Report) Part2() int {
	larger := 0
	windows := r.slidingWindow()
	for i := 0; i < len(windows)-1; i++ {
		if windows[i] < windows[i+1] {
			larger += 1
		}

	}
	return larger
}

func (r *Report) Part1() int {
	larger := 0
	for i := 0; i < len(r.depths)-1; i++ {
		if r.depths[i] < r.depths[i+1] {
			larger += 1
		}

	}
	return larger
}

func main() {
	report := &Report{}
	err := report.Parse(input)
	if err != nil {
		panic("Error parsing input: " + err.Error())
	}
	fmt.Printf("part1: %d", report.Part1())
	fmt.Printf("part2: %d", report.Part2())
}
